### PR TITLE
CS-99 Apprentice Commitments

### PR DESCRIPTION
**Pull request by benagain**
New outer API for the Apprentice Commitments project

### CODE REVIEW

File: src/SFA.DAS.ApprenticeCommitments.Api/AppStart/AddServiceRegistrationExtension.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    public static class AddServiceRegistrationExtension
+    {
+        public static void AddServiceRegistration(this IServiceCollection services)
+        {
+            services.AddHttpClient();
+            services.AddTransient<IAzureClientCredentialHelper, AzureClientCredentialHelper>();
+
+            services.AddTransient<ApprenticeLoginClient>();
+            services.AddTransient(typeof(IInternalApiClient<>), typeof(InternalApiClient<>));
+            services.AddTransient<IInternalApiClient<ApprenticeLoginConfiguration>>(s =>
```
Comment (dashton82) [inline src/SFA.DAS.ApprenticeCommitments.Api/AppStart/AddServiceRegistrationExtension.cs:21]: should this be:
```
services.AddTransient<IAprenticeLoginClient<CoursesApiConfiguration>, AprenticeLoginClient>();
```

File: src/SFA.DAS.ApprenticeCommitments/Apis/CommitmentsV2InnerApi/ApprenticeshipResponse.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        public string CourseName { get; set; }
+        public DateTime StartDate { get; set; }
+        public DateTime EndDate { get; set; }
+        public DateTime DateOfBirth { get; set; }
+        public DateTime? StopDate { get; set; }
+        public DateTime? PauseDate { get; set; }
+        public DateTime? CompletionDate { get; set; }
+        public bool HasHadDataLockSuccess { get; set; }
+        public DateTime? OriginalStartDate { get; set; }
+        //public ApprenticeshipEmployerType? ApprenticeshipEmployerTypeOnApproval { get; set; }
```
Comment (dashton82) [inline src/SFA.DAS.ApprenticeCommitments/Apis/CommitmentsV2InnerApi/ApprenticeshipResponse.cs:23]: Need to be there or should be deleted?

File: src/SFA.DAS.ApprenticeCommitments/Apis/ApprenticeLoginApi/SendInvitationRequest.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
+﻿using System;
+using Newtonsoft.Json;
+using SFA.DAS.SharedOuterApi.Interfaces;
+
+namespace SFA.DAS.ApprenticeCommitments.Apis.ApprenticeLoginApi
+{
+    public class SendInvitationRequest : IPostApiRequest
+    {
+        public string PostUrl => $"/invitations/{ClientId}";
```
Comment (dashton82) [inline src/SFA.DAS.ApprenticeCommitments/Apis/ApprenticeLoginApi/SendInvitationRequest.cs:9]: Maybe have the `ClientId` as ctor param - similar to what is done for the `GetApprenticeshipDetailsRequest`

File: src/SFA.DAS.ApprenticeCommitments/Apis/GetHealthRequest.cs [hunk 4]
```diff
@@ -0,0 +1,9 @@
+using SFA.DAS.SharedOuterApi.Interfaces;
+
+namespace SFA.DAS.ApprenticeCommitments.Apis
+{
+    public class GetHealthRequest : IGetApiRequest
```
Comment (dashton82) [inline src/SFA.DAS.ApprenticeCommitments/Apis/GetHealthRequest.cs:5]: There is a `GetPingRequest` in shared that can be reused

File: src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            await _apprenticeLoginService.SendInvitation(new SendInvitationModel
+            {
+                SourceId = id,
+                Email = command.Email,
+                GivenName = apprentice.FirstName,
+                FamilyName = apprentice.LastName,
+                OrganisationName = command.Organisation,
+                ApprenticeshipName = apprentice.CourseName,
+            });
+
```
Comment (dashton82) [inline src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs:50]: Can Task.WhenAll be used instead for these async requests?

File: src/SFA.DAS.ApprenticeCommitments/Apis/CommitmentsV2InnerApi/ApprenticeshipResponse.cs (hunk 2 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Apis/CommitmentsV2InnerApi/ApprenticeshipResponse.cs:23]: Removed

File: src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs (hunk 5 above)
Comment (benagain) [reply src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs:50]: I did initially, but it makes more sense to have them sequential as the actions make increasingly larger visible changes to the system.  If one operation fails it is prefferable to refrain from performing the subsequent operation(s).

Comment (dashton82) [reply src/SFA.DAS.ApprenticeCommitments/Application/Commands/CreateApprenticeship/CreateApprenticeshipCommandHandler.cs:50]: 👍🏻

**Review by dashton82 — APPROVED** [review]

**Pull request merged**